k8s: keep CiliumNode labels synced with Node object

jibi opened the pull request:
See individual commits

aanm commented on daemon/cmd/daemon.go:
```diff
@@ -525,6 +525,8 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
 		d.k8sWatcher.NodeChain.Register(&d.k8sWatcher.K8sSvcCache)
 	}
 
+	d.k8sWatcher.NodeChain.Register((*watchers.CiliumNodeLabelsUpdater)(d.k8sWatcher))
```
`d.k8sWatcher.NodeChain.Register((*watchers.CiliumNodeLabelsUpdater)(d.k8sWatcher))` can be `d.k8sWatcher.NodeChain.Register(d.k8sWatcher)`. The code will compile as well.

aanm commented on pkg/k8s/watchers/cilium_node.go:
```diff
@@ -4,108 +4,85 @@
-		// Create a new node controller when we are disconnected with the
-		// kvstore
-		<-kvstore.Client().Disconnected()
+	k.ciliumNodeStoreMU.Lock()
```
There's no need to protect `k.ciliumNodeStore` with a lock if we guarantee that `k.ciliumNodeStore` is set before we execute `func (k *K8sWatcher) updateCiliumNodeLabels(node *v1.Node)`

jibi replied:
I don't think there's any guarantee that `k.ciliumNodeInit()` will not run concurrently with the `sync-node-labels-with-ciliumnode` `Node` controller, so we should keep protecting this

aanm reviewed (approved):
Just 2 small nits. LGTM, doesn't require my 2nd review. However, it would be great if we could have some e2e tests that covers this.

jibi committed:
k8s: keep CiliumNode labels synced with Node object

This commit ensures that any update to the labels of a k8s Node object
is reflected in the corresponding CiliumNode one.

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

jibi commented on daemon/cmd/daemon.go:
```diff
@@ -527,6 +527,8 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
 		d.k8sWatcher.NodeChain.Register(&d.k8sWatcher.K8sSvcCache)
 	}
 
+	d.k8sWatcher.NodeChain.Register(d.k8sWatcher)
```
I'm now wondering if it would be better to wrap `K8sWatcher` in a new type rather then aliasing it to make it more explicit what's going on here :thinking:

aanm replied:
Have you thought on simply implementing the methods of Node interface directly on the `K8sWatcher`? :thinking:

jibi replied:
Yes, but I think that would preclude us from adding additional implementations of the `subscriber.Node` interface in the future in case we need to. Also having a specific type (in this case `CiliumNodeLabelsUpdater`) should make it more clear what that implementation is doing.

In short I think we should either:
* get rid of the alias, as it doesn't give us any benefit, and implement it on the `K8sWatcher` (with the :point_up: caveat)
* replace the alias with a new type that wraps `K8sWatcher` (to be more explicit and future-proof)

No strong opinion here as both approaches work fine at the moment

aanm reviewed (changes requested):
Discussed offline, there are some necessary changes to handle the kvstore mode node update

ldelossa commented on pkg/k8s/watchers/node.go:
```diff
@@ -117,3 +123,67 @@ func (k *K8sWatcher) GetK8sNode(_ context.Context, nodeName string) (*v1.Node, e
 }
+
+// CiliumNodeLabelsUpdater implements the subscriber.Node interface and is used
+// to keep CiliumNode objects labels in sync with the node ones.
```
```suggestion
// to keep CiliumNode objects labels in sync with the node's labels
```

christarazi commented on pkg/k8s/watchers/node.go:
```diff
@@ -117,3 +135,109 @@ func (k *K8sWatcher) GetK8sNode(_ context.Context, nodeName string) (*v1.Node, e
+
+// CiliumNodeUpdater implements the subscriber.Node interface and is used
+// to keep CiliumNode objects in sync with the node ones.
+type CiliumNodeUpdater struct {
```
```suggestion
// ciliumNodeUpdater implements the subscriber.Node interface and is used
// to keep CiliumNode objects in sync with the node ones.
type ciliumNodeUpdater struct {
```

Doesn't need to be exported.

christarazi commented on pkg/k8s/watchers/node.go:
```diff
@@ -117,3 +135,109 @@ func (k *K8sWatcher) GetK8sNode(_ context.Context, nodeName string) (*v1.Node, e
+						})
+					}
+
+					for _, address := range nodeDiscovery.LocalNode.IPAddresses {
```
[1] 

I think we need to be careful here because `LocalNode` can be written to async from somewhere else. Especially with @gandro's [PR](https://github.com/cilium/cilium/pull/18887) where this variable will change at runtime, we'll need a way to retrieve state from it safely. Maybe it's worth defining an interface to abstract away the underlying object?

christarazi commented on pkg/k8s/watchers/node.go:
```diff
@@ -117,3 +135,109 @@ func (k *K8sWatcher) GetK8sNode(_ context.Context, nodeName string) (*v1.Node, e
+func NewCiliumNodeUpdater(k8sWatcher *K8sWatcher, nodeDiscovery *nodediscovery.NodeDiscovery) *CiliumNodeUpdater {
+	return &CiliumNodeUpdater{
+		k8sWatcher:    k8sWatcher,
+		nodeDiscovery: nodeDiscovery,
```
[2]

I'm wondering if it would be better to scope down the access of `nodeDiscovery` from within this object. Can we potentially find an interface to abstract this away? This can also ease unit tests.

jibi replied:
So `nodeDiscovery` is used to:
* access the KV store: `nodeDiscovery.Registrar.UpdateLocalKeySync(k8sNodeParsed)`
* access the local node addresses: `nodeDiscovery.LocalNode.IPAddresses`

For the first point I think we can simply define an interface like:
```
type KVStoreNodeUpdater interface {
       UpdateNode(node *nodeTypes.Node) error
}
```
(and have `NodeDiscovery` implement it).

Not sure about what we should expose for the second point (just the bare minimum i.e. a way to access `LocalNode`'s IP addresses in a safe way?)

christarazi commented on test/k8sT/Node.go:
```diff
@@ -0,0 +1,55 @@
+	"github.com/cilium/cilium/test/helpers"
+)
+
+var _ = Describe("K8sNode", func() {
```
Hm, do we really need an e2e test for this functionality? Couldn't we write unit tests to assert that `(u *CiliumNodeUpdater) updateCiliumNode` does the right thing? I don't think we need a full K8s cluster to validate the behavior of that function. Granted, that's assuming that we implement [1] and [2], as well as abstracting away the K8s client.

jibi replied:
I thought it would be a good way to test the full pipeline (update Node -> node watcher -> controller -> update CiliumNode -> read new object from API server), but if we improve unit tests we could just get rid of this and still test it indirectly in the egress gw e2e tests (where an update to a Node should trigger the reconciliation logic)

christarazi reviewed (changes requested):
Thanks for the PR!

I have a couple comments about how we can ease testing and concerns about accessing the nodediscovery package from within the controller.

gandro commented on pkg/k8s/watchers/node.go:
```diff
@@ -117,3 +135,104 @@ func (k *K8sWatcher) GetK8sNode(_ context.Context, nodeName string) (*v1.Node, e
+}
+
+func (u *ciliumNodeUpdater) OnUpdateNode(oldNode, newNode *v1.Node, swg *lock.StoppableWaitGroup) error {
+	u.updateCiliumNode(u.kvStoreNodeUpdater, u.localNode, newNode)
```
~Maybe I'm missing something, but do you not need to check that `newNode` actually refers to the local node? This function will be called if any CiliumNode object is updated, and I don't think we want the local cilium-agent to update labels of a CiliumNode object of a remote node?~

Edit: Sorry. Nevermind. I confused `OnUpdateNode` with `OnUpdateCiliumNode`. The former is only called for the local node, only the latter is called for every node.

jibi replied:
Correct, the agent is supposed to update labels and addresses only of the CiliumNode object for its local node, and this is guaranteed by the fact that the `Node` watcher (which invokes this logic) is only subscribed to events for its own node:

https://github.com/cilium/cilium/blob/f2ecb1aa8ba2a86d49b6288dc25842f2eabcd453/pkg/k8s/watchers/node.go#L57-L62

christarazi commented on pkg/k8s/watchers/node.go:
```diff
@@ -117,3 +135,104 @@ func (k *K8sWatcher) GetK8sNode(_ context.Context, nodeName string) (*v1.Node, e
+		kvStoreNodeUpdater: kvStoreNodeUpdater,
+		localNode:          localNode,
+	}
+}
```
Ideally, the interfaces are defined where they are being used (within this file), but given that we've discussed a refactor after this is merged offline anyway, we can address this then.

jibi replied:
Interfaces moved to `pkg/k8s/watchers/node.go` :+1:

jibi committed:
watchers: don't ignore CiliumNode events if labels get updated

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

qmonnet commented on test/k8s/Node.go:
```diff
@@ -0,0 +1,58 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2021 Authors of Cilium
```
```suggestion
// Copyright Authors of Cilium
```

qmonnet replied:
I thought you'd fixed it! :scream_cat:

jibi replied:
Crap :disappointed: 
I actually fixed it initially, but then I must have screwed up with git and lost that change

qmonnet replied:
Never mind :), I'll do another pass someday for the new files people were working on when we pushed the change.

aanm commented on pkg/k8s/watchers/node.go:
```diff
@@ -37,6 +46,18 @@ func (k *K8sWatcher) RegisterNodeSubscriber(s subscriber.Node) {
+	if !reflect.DeepEqual(oldNode.Status.Addresses, newNode.Status.Addresses) {
+		return false
+	}
+
```
If the title says we won't do anything for addresses then we can drop this?

jibi replied:
Correct, this is not needed now, I'll drop it :+1: 
(but I'll still keep this function as I'll probably need the annotations check for the KV store case)
